### ISSUE TITLE
The OCaml bindings heavily use naked pointers, which have been deprecated in OCaml 4.11 and will not be supported in OCaml 5

### ISSUE THREAD

**Issue by alan-j-hu** [llvm/llvm-project#58134]
When modifying the LLVM OCaml bindings, I noticed that the code uses naked pointers (pointers to memory outside the OCaml heap that are not wrapped in a special block). For just one example, see

https://github.com/llvm/llvm-project/blob/0a9756fc15c57dc94146e65d251d7a3d941ae78a/llvm/bindings/ocaml/llvm/llvm_ocaml.c#L203

An `LLVMContextRef` is directly returned.

[Since OCaml 4.11, naked pointers have been deprecated. They will not be supported in OCaml 5](https://v2.ocaml.org/manual/intfc.html#ss:c-outside-head), which will move to a new runtime. The recommended solution is to either wrap the pointer in an OCaml block (heap allocation) with an `Abstract_tag` or `Custom_tag`; the GC will not scan the elements of these blocks.

Furthermore, the LLVM bindings contain many resource cleanup functions, such as https://github.com/llvm/llvm-project/blob/0a9756fc15c57dc94146e65d251d7a3d941ae78a/llvm/bindings/ocaml/llvm/llvm.mli#L454 among many others. OCaml blocks with the `Custom_tag` [support finalization functions](https://v2.ocaml.org/manual/intfc.html#s%3Ac-custom). It would be worth it to make use of this feature if removing naked pointers.

I am willing to rewrite the OCaml bindings to use blocks instead of naked pointers, but this is a rather big task, so I wanted to post an issue first and hear the input of LLVM experts. I know that the OCaml OPAM maintainers manually track package build status, so they might also be interested.

See also https://discuss.ocaml.org/t/ann-a-dynamic-checker-for-detecting-naked-pointers/5805

**Comment by alan-j-hu** [comment]
For what it's worth, when I switched to the `ocaml-variants.4.10.0+nnpcheck` compiler switch and built `check-llvm-bindings-ocaml`, I got:

```
[15/16] Running lit suite /home/alan/llvm-project/llvm/test/Bindings/OCaml
llvm-lit: /home/alan/llvm-project/llvm/utils/lit/lit/llvm/subst.py:126: note: Did not find llvm-debuginfod in /home/alan/llvm-project/build/bin
FAIL: LLVM :: Bindings/OCaml/core.ml (1 of 16)
******************** TEST 'LLVM :: Bindings/OCaml/core.ml' FAILED ********************
Script:
--
: 'RUN: at line 1';   rm -rf /home/alan/llvm-project/build/test/Bindings/OCaml/Output/core.ml.tmp && mkdir -p /home/alan/llvm-project/build/test/Bindings/OCaml/Output/core.ml.tmp && cp /home/alan/llvm-project/llvm/test/Bindings/OCaml/core.ml /home/alan/llvm-project/build/test/Bindings/OCaml/Output/core.ml.tmp/core.ml && cp /home/alan/llvm-project/llvm/test/Bindings/OCaml/Utils/Testsuite.ml /home/alan/llvm-project/build/test/Bindings/OCaml/Output/core.ml.tmp/Testsuite.ml
: 'RUN: at line 2';   /home/alan/.opam/4.10.0+nnpcheck/bin/ocamlfind ocamlc -cclib -L/home/alan/llvm-project/build/lib  -g -w +A -package llvm.analysis -package llvm.bitwriter -I /home/alan/llvm-project/build/test/Bindings/OCaml/Output/core.ml.tmp/ -linkpkg /home/alan/llvm-project/build/test/Bindings/OCaml/Output/core.ml.tmp/Testsuite.ml /home/alan/llvm-project/build/test/Bindings/OCaml/Output/core.ml.tmp/core.ml -o /home/alan/llvm-project/build/test/Bindings/OCaml/Output/core.ml.tmp/executable
: 'RUN: at line 3';   /home/alan/llvm-project/build/test/Bindings/OCaml/Output/core.ml.tmp/executable /home/alan/llvm-project/build/test/Bindings/OCaml/Output/core.ml.tmp/bitcode.bc
: 'RUN: at line 4';   /home/alan/.opam/4.10.0+nnpcheck/bin/ocamlfind ocamlopt -cclib -L/home/alan/llvm-project/build/lib -cclib -Wl,-rpath,/home/alan/llvm-project/build/lib  -g -w +A -package llvm.analysis -package llvm.bitwriter -I /home/alan/llvm-project/build/test/Bindings/OCaml/Output/core.ml.tmp/ -linkpkg /home/alan/llvm-project/build/test/Bindings/OCaml/Output/core.ml.tmp/Testsuite.ml /home/alan/llvm-project/build/test/Bindings/OCaml/Output/core.ml.tmp/core.ml -o /home/alan/llvm-project/build/test/Bindings/OCaml/Output/core.ml.tmp/executable
: 'RUN: at line 5';   /home/alan/llvm-project/build/test/Bindings/OCaml/Output/core.ml.tmp/executable /home/alan/llvm-project/build/test/Bindings/OCaml/Output/core.ml.tmp/bitcode.bc
: 'RUN: at line 6';   /home/alan/llvm-project/build/bin/llvm-dis < /home/alan/llvm-project/build/test/Bindings/OCaml/Output/core.ml.tmp/bitcode.bc > /home/alan/llvm-project/build/test/Bindings/OCaml/Output/core.ml.tmp/dis.ll
: 'RUN: at line 7';   /home/alan/llvm-project/build/bin/FileCheck /home/alan/llvm-project/llvm/test/Bindings/OCaml/core.ml < /home/alan/llvm-project/build/test/Bindings/OCaml/Output/core.ml.tmp/dis.ll
: 'RUN: at line 9';   /home/alan/llvm-project/build/bin/FileCheck -check-prefix=CHECK-NOWHERE /home/alan/llvm-project/llvm/test/Bindings/OCaml/core.ml < /home/alan/llvm-project/build/test/Bindings/OCaml/Output/core.ml.tmp/dis.ll
--
Exit Code: 139

Command Output (stderr):
--
/home/alan/llvm-project/build/test/Bindings/OCaml/Output/core.ml.script: line 8: 37261 Segmentation fault      (core dumped) /home/alan/llvm-project/build/test/Bindings/OCaml/Output/core.ml.tmp/executable /home/alan/llvm-project/build/test/Bindings/OCaml/Output/core.ml.tmp/bitcode.bc

--

********************
FAIL: LLVM :: Bindings/OCaml/debuginfo.ml (4 of 16)
******************** TEST 'LLVM :: Bindings/OCaml/debuginfo.ml' FAILED ********************
Script:
--
: 'RUN: at line 1';   rm -rf /home/alan/llvm-project/build/test/Bindings/OCaml/Output/debuginfo.ml.tmp && mkdir -p /home/alan/llvm-project/build/test/Bindings/OCaml/Output/debuginfo.ml.tmp && cp /home/alan/llvm-project/llvm/test/Bindings/OCaml/debuginfo.ml /home/alan/llvm-project/build/test/Bindings/OCaml/Output/debuginfo.ml.tmp/debuginfo.ml && cp /home/alan/llvm-project/llvm/test/Bindings/OCaml/Utils/Testsuite.ml /home/alan/llvm-project/build/test/Bindings/OCaml/Output/debuginfo.ml.tmp/Testsuite.ml
: 'RUN: at line 2';   /home/alan/.opam/4.10.0+nnpcheck/bin/ocamlfind ocamlc -cclib -L/home/alan/llvm-project/build/lib  -g -w +A -package llvm.all_backends -package llvm.target -package llvm.analysis -package llvm.debuginfo -I /home/alan/llvm-project/build/test/Bindings/OCaml/Output/debuginfo.ml.tmp/ -linkpkg /home/alan/llvm-project/build/test/Bindings/OCaml/Output/debuginfo.ml.tmp/Testsuite.ml /home/alan/llvm-project/build/test/Bindings/OCaml/Output/debuginfo.ml.tmp/debuginfo.ml -o /home/alan/llvm-project/build/test/Bindings/OCaml/Output/debuginfo.ml.tmp/executable
: 'RUN: at line 3';   /home/alan/llvm-project/build/test/Bindings/OCaml/Output/debuginfo.ml.tmp/executable | /home/alan/llvm-project/build/bin/FileCheck /home/alan/llvm-project/llvm/test/Bindings/OCaml/debuginfo.ml
: 'RUN: at line 4';   /home/alan/.opam/4.10.0+nnpcheck/bin/ocamlfind ocamlopt -cclib -L/home/alan/llvm-project/build/lib -cclib -Wl,-rpath,/home/alan/llvm-project/build/lib  -g -w +A -package llvm.all_backends -package llvm.target -package llvm.analysis -package llvm.debuginfo -I /home/alan/llvm-project/build/test/Bindings/OCaml/Output/debuginfo.ml.tmp/ -linkpkg /home/alan/llvm-project/build/test/Bindings/OCaml/Output/debuginfo.ml.tmp/Testsuite.ml /home/alan/llvm-project/build/test/Bindings/OCaml/Output/debuginfo.ml.tmp/debuginfo.ml -o /home/alan/llvm-project/build/test/Bindings/OCaml/Output/debuginfo.ml.tmp/executable
: 'RUN: at line 5';   /home/alan/llvm-project/build/test/Bindings/OCaml/Output/debuginfo.ml.tmp/executable | /home/alan/llvm-project/build/bin/FileCheck /home/alan/llvm-project/llvm/test/Bindings/OCaml/debuginfo.ml
--
Exit Code: 2

Command Output (stderr):
--
FileCheck error: '<stdin>' is empty.
FileCheck command line:  /home/alan/llvm-project/build/bin/FileCheck /home/alan/llvm-project/llvm/test/Bindings/OCaml/debuginfo.ml

--

********************
********************
Failed Tests (2):
  LLVM :: Bindings/OCaml/core.ml
  LLVM :: Bindings/OCaml/debuginfo.ml


Testing Time: 20.40s
  Passed: 14
  Failed:  2
FAILED: test/CMakeFiles/check-llvm-bindings-ocaml /home/alan/llvm-project/build/test/CMakeFiles/check-llvm-bindings-ocaml 
cd /home/alan/llvm-project/build/test && /usr/bin/python3.10 /home/alan/llvm-project/build/./bin/llvm-lit -sv /home/alan/llvm-project/llvm/test/Bindings/OCaml
ninja: build stopped: subcommand failed.
```

I just got a change to the OCaml bindings committed a few days ago, and I never got this error when testing, so I assume that switching to the naked pointer detector switch caused these test failures. When testing my previous changes, I used the plain OCaml 4.14 compiler switch with opam.

I will do more investigation.

**Comment by alan-j-hu** [comment]
After taking a closer look, I realized that using OCaml's finalization feature is actually a bad idea. The reason is that LLVM objects may have pointers to each other (e.g. an `lltype` has an `llcontext`), but these relationships are of course invisible to the OCaml GC since it cannot follow pointers outside the OCaml heap. If OCaml's finalization feature were to be used, the GC might finalize an LLVM object that is still reachable from some still live LLVM object. Therefore, I think it's best to stick to `Abstract_tag` blocks, not `Custom_tag` blocks with finalizers, and to keep manual cleanup functions.

**Comment by jberdine** [comment]
> The recommended solution is to either wrap the pointer in an OCaml block (heap allocation) with an Abstract_tag or Custom_tag; the GC will not scan the elements of these blocks.

If you are not going to set custom functions for comparison, hashing, or finalization, then I think that it might be easier / less code to wrap pointers to llvm memory using the ocaml Nativeint, rather than manually using Abstract_tag or Custom_tag objects.

It might be that all the pointers returned from llvm to ocaml are at least 2-byte aligned. If so, then another potential implementation approach is to set the low bit of pointers when returning them from llvm to ocaml, and to clear the low bit when passing them back from ocaml to llvm. Setting the low bit will make the encoded pointer appear to be an integer to the ocaml gc, and will not be scanned. The advantage would be that it would save allocating the wrapper objects. I am not sure how to validate the alignment requirements though. It would be possible to add some assertions to check the alignment, and run some tests, but that would provide only partial evidence.

A hybrid approach is also likely possible, where the conversion from llvm to ocaml tests the low bit, sets it if clear, and allocates a wrapper object if not. Then the conversion from ocaml to llvm could test the low bit and either clear it if set, or treat it as a pointer to a wrapper object if clear. All objects allocated by ocaml will be word-aligned, so testing the low bit will suffice to distinguish the two cases.

**Comment by alan-j-hu** [comment]
In terms of either verbosity or efficiency, I don't think Nativeint or Abstract_tag have an advantage over each other. I personally think that using `Abstract_tag` makes the code's intent clearer.

Treating pointers outside the OCaml heap as OCaml integers sounds appealing to me because it avoids the allocation. I was reluctant to do this because it depended on an aspect of the LLVM type's representation that was not part of the API, but as the OCaml bindings are an official part of the LLVM project, I figure that the OCaml bindings would have a guarantee not to suddenly break due to implementation changes elsewhere.

Could we check alignment just by reading the type definitions? I couldn't find definitions of `struct LLVMOpaqueType`, `struct LLVMOpaqueValue`, etc, that `LLVMTypeRef`, `LLVMValueRef` are [`typedef`ed](https://llvm.org/doxygen/group__LLVMCSupportTypes.html) as pointers to, so I'm assuming that they don't have definitions and are just used to cast from pointers to `llvm::Type`, `llvm::Value`, etc. If these types are parent classes in a hierarchy, checking all the subclasses would be pain, but the objects will carry around a pointer to a vtable, so they will have at least a word alignment, unless I am misunderstanding something? And is there a chance that there will be a weird platform where say, an `int` has 1-byte alignment, that the bindings should still be able to compile and run on?

Alternatively, I was wondering if the alignment of a type could be programmatically checked at compile-time, and the proper wrapping/unwrapping functions selected? I thought that maybe `Alignof_` could work, but unfortunately, it can't be checked from a C macro. However, if a check on a type's alignment were to be inserted into an if/else statement, because this information is known at compile-time, the dead branch could easily be optimized out by the compiler. `Alignof_` requires C11. Is there a specific minimum version of C that the bindings are required to compile with?

An advantage of the hybrid approach over a rigid approach that makes the decision on a per-type basis is that while the latter will always allocate if the type has 1-byte alignment, the hybrid approach will still avoid the allocation if the specific pointer is even, even if the type itself is 1-byte aligned. I think this approach is also the most "OCamly." OCaml uses a similar technique where within a variant type with both nullary data constructors and data constructors carrying values, the former are unboxed and represented as odd integers, while the latter are boxed and carry a tag.

Do you think that making sure that all pointers of a specific type are not 1-byte aligned is too error-prone, or do you feel confident that this can be implemented and maintained?

I personally think that your hybrid approach is the best: simple while still avoiding unnecessary allocations.

**Comment by alan-j-hu** [comment]
If the bindings are going to be rewritten, what do you think of using https://github.com/yallop/ocaml-ctypes? The bindings seemingly require ctypes and ctypes-foreign, but I'm not sure where they're used...

Updating the bindings is tedious, but it should mostly be a one-time pain. On the other hand, ctypes is a little too complicated for me to understand if I need to debug how code is being generated, but from what I can tell, it doesn't make the optimization you suggest about not boxing even pointers. I am worried that ctypes won't give enough control over the generated code.

**Comment by jberdine** [comment]
I don't have enough personal experience with ctypes to advise one way or the other. One point to keep in mind though is that it would be difficult to do if it requires adding build or test time dependencies. Most LLVM developers, and the build bots, are not going to have additional dependencies.

**Comment by jberdine** [comment]
Fixed by 30f423a543641816e367cd17dad3d4808cba8a79, thanks @alan-j-hu!
